### PR TITLE
PIP-25: C++ / Python / Go implementation for token authentication

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Added C++ client implementation along with Python and Go bindings for token based authentication.

Note: tests are still failing since will need #2888 to be merged.

### CODE REVIEW

File: pulsar-client-cpp/python/pulsar_test.py [hunk 1]
```diff
@@ -793,6 +793,43 @@ def test_get_topics_partitions(self):
...
+        self.assertTrue(msg)
+        self.assertEqual(msg.data(), b'hello')
+        client.close()
+
+    def test_token_auth_supplier(self):
+        def read_token():
+            with open('/tmp/pulsar-test-data/certs/token.txt') as tf:
+                return tf.read().strip()
+
+        client = Client(self.serviceUrl,
```
Comment (ivankelly) [inline pulsar-client-cpp/python/pulsar_test.py]: why serviceUrl and http above? If you use http here it will still exercise both endpoints

File: pulsar-client-cpp/python/src/authentication.cc [hunk 2]
```diff
@@ -32,6 +32,59 @@ struct AuthenticationTlsWrapper : public AuthenticationWrapper {
...
+        Py_XINCREF(_pySupplier);
+        return *this;
+    }
+
+    virtual ~TokenSupplierWrapper() {
+        Py_XDECREF(_pySupplier);
+    }
+
+    std::string operator()() {
+        PyGILState_STATE state = PyGILState_Ensure();
```
Comment (ivankelly) [inline pulsar-client-cpp/python/src/authentication.cc]: why do you need to acquire the GIL? _pySupplier is immutable.

File: pulsar-client-cpp/python/src/authentication.cc [hunk 3]
```diff
@@ -32,6 +32,59 @@ struct AuthenticationTlsWrapper : public AuthenticationWrapper {
     }
 };
 
+struct TokenSupplierWrapper {
+    PyObject* _pySupplier;
```
Comment (ivankelly) [inline pulsar-client-cpp/python/src/authentication.cc]: const ptr

File: pulsar-client-cpp/tests/AuthTokenTest.cc [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        std::string expectedMessageContent = prefix + boost::lexical_cast<std::string>(i);
+        LOG_INFO("Received Message with [ content - "
+                 << receivedMsg.getDataAsString() << "] [ messageID = " << receivedMsg.getMessageId() << "]");
+        ASSERT_EQ(receivedMsg.getProperty("msgIndex"), boost::lexical_cast<std::string>(i));
+        ASSERT_EQ(expectedMessageContent, receivedMsg.getDataAsString());
+        ASSERT_EQ(ResultOk, consumer.acknowledge(receivedMsg));
+    }
+}
+
+TEST(AuthPluginToken, testTokenWithHttpUrl) {
```
Comment (ivankelly) [inline pulsar-client-cpp/tests/AuthTokenTest.cc]: This exercises the binary path too. I don't think you need both.

File: pulsar-client-cpp/tests/CMakeLists.txt [hunk 5]
```diff
@@ -25,4 +25,4 @@ add_executable(main ${TEST_SOURCES})
 
 target_include_directories(main PRIVATE ${CMAKE_SOURCE_DIR}/lib)
 
-target_link_libraries(main ${CLIENT_LIBS} ${GMOCK_LIBRARY_PATH})
+target_link_libraries(main ${CLIENT_LIBS} pulsarShared ${GMOCK_LIBRARY_PATH})
```
Comment (ivankelly) [inline pulsar-client-cpp/tests/CMakeLists.txt]: seems unrelated.

File: pulsar-client-go/pulsar/c_client.go [hunk 6]
```diff
@@ -127,6 +127,37 @@ func newAuthenticationTLS(certificatePath string, privateKeyPath string) Authent
 	return auth
```
Comment (ivankelly) [inline pulsar-client-go/pulsar/c_client.go]: I haven't reviewed this, don't know go.

File: pulsar-client-cpp/include/pulsar/Authentication.h [hunk 7]
```diff
@@ -114,6 +115,43 @@ class AuthTls : public Authentication {
     AuthenticationDataPtr authDataTls_;
 };
 
+typedef boost::function<std::string()> TokenSupplier;
```
[the remark this answers is not in the document]
Comment (maskit) [reply pulsar-client-cpp/include/pulsar/Authentication.h]: Also, we can use `using` instead of `typedef` in C++11.

File: pulsar-client-cpp/lib/auth/AuthToken.cc [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/bind.hpp>
+
+#include <fstream>
+
+namespace pulsar {
+
+// AuthDataToken
+
+AuthDataToken::AuthDataToken(const TokenSupplier &tokenSupplier) { tokenSupplier_ = tokenSupplier; }
```
Comment (maskit) [inline pulsar-client-cpp/lib/auth/AuthToken.cc]: I use initializer list whenever it's possible.
https://www.geeksforgeeks.org/when-do-we-use-initializer-list-in-c/

btw, is it for security that this class doesn't have an actual token as a member? (to not keep a raw token on memory long time?)

File: pulsar-client-cpp/lib/c/c_Authentication.cc [hunk 9]
```diff
@@ -23,6 +23,8 @@
 
 #include "c_structs.h"
 
+#include <stdlib.h>
```
Comment (maskit) [inline pulsar-client-cpp/lib/c/c_Authentication.cc]: cstdlib?

File: pulsar-client-cpp/lib/auth/AuthToken.cc (hunk 8 above)
Comment (maskit) [reply pulsar-client-cpp/lib/auth/AuthToken.cc]: oh, I see. TokenSupplier needs to receive a string.

File: pulsar-client-cpp/python/src/authentication.cc (hunk 2 above)
Comment (merlimat) [reply pulsar-client-cpp/python/src/authentication.cc]: We need the GIL because we're calling into python code (supplied by the user)

File: pulsar-client-cpp/include/pulsar/Authentication.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (merlimat) [reply pulsar-client-cpp/include/pulsar/Authentication.h]: Throughout the API we are using `boost::function` (as well as `boost::shared_ptr`). This comes from the fact that code was initially written as C++98 compliant.

I believe we should move to C++11 standard (actually C++14 at this point) and convert all these boost references into std, in one shot. 

Regarding this change, I believe we should be consistent with the rest of API.

[the remark this answers is not in the document]
Comment (ivankelly) [reply pulsar-client-cpp/include/pulsar/Authentication.h]: works for me. Can we actually move to c++11 (or 14 even)? I remember when I was in yahoo, the redhat version, and consequently the stdc++ version, were as old as the hills. I don't know if that's still the case. In any case, C++11 should be generally available by now.

File: pulsar-client-cpp/python/src/authentication.cc (hunk 2 above)
Comment (ivankelly) [reply pulsar-client-cpp/python/src/authentication.cc]: :+1:

File: pulsar-client-cpp/include/pulsar/Authentication.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (merlimat) [reply pulsar-client-cpp/include/pulsar/Authentication.h]: I C++11 is fine with GCC-4.8 (that ships with RHEL7 - https://access.redhat.com/solutions/19458), though C++14 would require GCC-5 (https://gcc.gnu.org/projects/cxx-status.html#cxx14) and might not as available

File: pulsar-client-cpp/tests/CMakeLists.txt (hunk 5 above)
Comment (merlimat) [reply pulsar-client-cpp/tests/CMakeLists.txt]: Yes, this is slightly unrelated, moving to its own PR.

File: pulsar-client-cpp/pulsar-test-service-start.sh [hunk 10]
```diff
@@ -44,6 +44,14 @@ mkdir -p $DATA_DIR
 mkdir -p $DATA_DIR/certs
 cp $SRC_DIR/pulsar-broker/src/test/resources/authentication/tls/*.pem $DATA_DIR/certs
 
+# Generate secret key and token
+$PULSAR_DIR/bin/pulsar tokens create-secret-key --output $DATA_DIR/certs/secret.key
```
[the remark this answers is not in the document]
Comment (merlimat) [reply pulsar-client-cpp/pulsar-test-service-start.sh]: Done

File: pulsar-client-cpp/python/pulsar_test.py [hunk 11]
```diff
@@ -793,6 +793,43 @@ def test_get_topics_partitions(self):
                          [topic_non_partitioned])
         client.close()
 
+    def test_token_auth(self):
+        with open('/tmp/pulsar-test-data/certs/token.txt') as tf:
+            token = tf.read().strip()
+        client = Client(self.adminUrl,
```
[the remark this answers is not in the document]
Comment (merlimat) [reply pulsar-client-cpp/python/pulsar_test.py]: Done

File: pulsar-client-cpp/tests/AuthTokenTest.cc (hunk 4 above)
Comment (merlimat) [reply pulsar-client-cpp/tests/AuthTokenTest.cc]: It's almost the same, but I wanted to make sure to cover the 2 possible ways to do the lookups.

File: pulsar-client-cpp/lib/c/c_Authentication.cc (hunk 9 above)
Comment (merlimat) [reply pulsar-client-cpp/lib/c/c_Authentication.cc]: Fixed

File: pulsar-client-cpp/python/src/authentication.cc (hunk 3 above)
Comment (merlimat) [reply pulsar-client-cpp/python/src/authentication.cc]: `boost::python::call()` below requires a non-const ` PyObject*`

**Comment by merlimat** [comment]
run java8 tests
run cpp tests
